firedrake_da/kalman/kalman.py

```diff
@@ -32,66 +35,101 @@ def Kalman_update(ensemble, observation_operator, observation_coords, observatio
-        D.append(f)
-
-    # now put this difference in matrix of data
-    d = np.zeros(np.shape(in_ensemble_dat))
-    for i in range(n):
-        d[:, i] = D[i].dat.data
-
-    """ kalman gain and update """
-    # make R matrix on observation space - independent observation error!
-    r = np.identity(len(d[:, 0])) * sigma
-
-    # kalman gain - NB: because all components of K are in observation space, operator H becomes id
-    K = np.dot(cov, np.linalg.inv(cov + r))
-    X = in_ensemble_dat + np.dot(K, d)
-
-    # formula for kalman gain
-    new_ensemble = []
-    for i in range(n):
-        f = Function(fs_to_project_to)
-        f.dat.data[:] = X[:, i]
-        new_ensemble.append(f)
-
-    """ project new ensemble back to original space """
-    transformed_ensemble = []
-
-    for i in range(n):
-        out_func = Function(original_space).project(new_ensemble[i])
-        transformed_ensemble.append(out_func)
-
-    return transformed_ensemble
+    with timed_stage("Initial observation instance"):
+        observation_operator.update_observation_operator(observation_coords, observations)
+    with timed_stage("Preallocating functions"):
+        diff = Function(vfsn)
+    with timed_stage("Calculating observation differences"):
+        if n == 1:
+            f = observation_operator.difference(ensemble[0], p)
+            diff.dat.data[:] = f.dat.data[:]
+        else:
+            for i in range(n):
+                f = observation_operator.difference(ensemble[i], p)
+                diff.dat.data[:, i] = f.dat.data[:]
+
+    # preallocate new ensemble function
+    with timed_stage("Preallocating functions"):
+        new_ensemble_f = Function(vfsn)
+
+    # now carry out kernel on multiplying kalman gain to differences
+    with timed_stage("Preallocating functions"):
+        increments = Function(vfsn)
+        tensor_increments = Function(tfs)
+
+    # take the transpose of kalman gain matrix
+    kalman_gain.dat.data[:] = np.matrix.transpose(np.reshape(kalman_gain.dat.data[:],
+                                                             ((nc, nc))))
+
+    # generate kernel and dictionary
+    with timed_stage("Carrying out kalman update"):
+        k_kernel = kalman_kernel_generation(n, nc)
+        Dict = {}
+        Dict.update({"product_tensor": (tensor_increments, WRITE)})
+        Dict.update({"input_matrix": (kalman_gain, READ)})
+        Dict.update({"input_vector": (diff, READ)})
+        par_loop(k_kernel.kalman_kernel, dx, Dict)
+
+        increments.dat.data[:] = np.sum(tensor_increments.dat.data, axis=0)
+
+    # add functions to find new ensemble
+    with timed_stage("Carrying out kalman update"):
+        new_ensemble_f = assemble(ensemble_f + increments)
+
+    # put back into individual functions
+    if n == 1:
+        ensemble[0].dat.data[:] = new_ensemble_f.dat.data[:]
+    else:
+        for i in range(n):
+            ensemble[i].dat.data[:] = new_ensemble_f.dat.data[:, i]
+
```

Comment:
This overrides the old ensemble array which is probably not suitable. Try to build a new preallocated new ensemble array - we can always make this change to this and `ensemble_transform` later if necessary (speed-up)